Encryption.md

```diff
@@ -19,216 +19,544 @@
+| Data Page            |     V      |    V (2)    |         V         |       V        |      V      |
+| Dictionary Page      |     V      |    V (3)    |         V         |       V        |      X      |
+| Data PageHeader      |     V      |    V (4)    |         V         |       V        |      V      |
+| Dictionary PageHeader|     V      |    V (5)    |         V         |       V        |      X      |
+| ColumnIndex          |     V      |    V (6)    |         V         |       V        |      X      |
+| OffsetIndex          |     V      |    V (7)    |         V         |       V        |      X      |
+
+
+
+## 5. File Format
+Parquet file encryption algorithm is specified in a union of the following Thrift structures:
+
+```c
+struct AesGcmV1 {
+  /** Retrieval metadata of AAD prefix **/
+  1: optional binary aad_metadata
+
+  /** Unique file identifier part of AAD suffix **/
+  2: optional binary aad_file_unique
+}
+
+struct AesGcmCtrV1 {
+  /** Retrieval metadata of AAD prefix **/
+  1: optional binary aad_metadata
+
+  /** Unique file identifier part of AAD suffix **/
+  2: optional binary aad_file_unique
+}
+
+union EncryptionAlgorithm {
+  1: AesGcmV1 AES_GCM_V1
+  2: AesGcmCtrV1 AES_GCM_CTR_V1
+}
+```
+ 
+
+The Thrift modules are encrypted after serialization, using the GCM cipher. For each module, the encryption 
+buffer is comprised of a nonce, ciphertext and tag, described in the Algorithms section. The column pages 
+(data and dictionary) are encrypted after compression. With the AES_GCM_V1 algorithm, a page encryption 
+buffer is comprised of a nonce, ciphertext and tag. The length of the encryption buffer 
+(a 4-byte little endian) is written to the output stream, followed by the buffer itself.
+
+|length (4 bytes) | nonce (12 bytes) | ciphertext (length-28 bytes) | tag (16 bytes) |
+|-----------------|------------------|------------------------------|----------------|
+
+
+With the AES_GCM_CTR_V1 algorithm, a page encryption buffer is comprised of a nonce and ciphertext, 
+described in the Algorithms section. The length of the encryption buffer 
+(a 4-byte little endian) is written to the output stream, followed by the buffer itself.
+
+|length (4 bytes) | nonce (12 bytes) | ciphertext (length-12 bytes) |
+|-----------------|------------------|------------------------------|
+
+
+A `crypto_meta_data` field is set in each ColumnChunk in the encrypted columns. ColumnCryptoMetaData 
+is a union - the actual structure is chosen depending on whether the column is encrypted with the 
+footer key, or with a column-specific key. For the latter, a key metadata can be specified.
+
+```c
+struct EncryptionWithFooterKey {
+}
+
+struct EncryptionWithColumnKey {
+  /** Column path in schema **/
+  1: required list<string> path_in_schema
+  
+  /** Retrieval metadata of the column-specific key **/
+  2: optional binary column_key_metadata
```

Comment:
Should this be named "key_metadata" since it is already in a column-specific structure?

Reply:
Same as above, 
https://github.com/apache/parquet-cpp/blob/apache-parquet-cpp-1.5.0/src/parquet/parquet.thrift#L672